BBQPlan should not be an EntryPlan

kmiller68 opened the pull request:
#### 754ed19097422f6e5912cb4cfb33afcf21245f16
<pre>
BBQPlan should not be an EntryPlan
<a href="https://bugs.webkit.org/show_bug.cgi?id=280274">https://bugs.webkit.org/show_bug.cgi?id=280274</a>
<a href="https://rdar.apple.com/136594153">rdar://136594153</a>

Reviewed by Yusuke Suzuki.

Right now BBQPlan is also an EntryPlan which means it needs to do all the
validation/module call graph steps required for entry. This was error prone
as e.g. BBQPlan with useWasmLLInt=0 would not report it&apos;s callsites to the
CalleeGroup thus would never would call OMG code.

This patch instead makes useWasmLLInt/useWasmIPInt=0 instantly trigger a
sync tier up to BBQ. This should simplify the code paths and make things
much simpler.

* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::BBQPlan):
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::compileFunction):
(JSC::Wasm::BBQPlan::prepareImpl): Deleted.
(JSC::Wasm::BBQPlan::didCompleteCompilation): Deleted.
(JSC::Wasm::BBQPlan::initializeCallees): Deleted.
(JSC::Wasm::BBQPlan::didReceiveFunctionData): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::jitCompileAndSetHeuristics):
* Source/JavaScriptCore/wasm/WasmIPIntTierUpCounter.h:
(JSC::Wasm::IPIntTierUpCounter::optimizeAfterWarmUp):
* Source/JavaScriptCore/wasm/WasmLLIntTierUpCounter.h:
(JSC::Wasm::LLIntTierUpCounter::optimizeAfterWarmUp):
* Source/JavaScriptCore/wasm/WasmModule.cpp:
(JSC::Wasm::Module::getOrCreateCalleeGroup):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::jitCompileAndSetHeuristics):
(JSC::LLInt::jitCompileSIMDFunction):

Canonical link: <a href="https://commits.webkit.org/284175@main">https://commits.webkit.org/284175@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e03e84bf58938505908dbdee0773c7de3b5f92e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68662 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48054 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21321 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72731 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19807 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55850 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19623 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13166 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71729 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59287 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35212 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16711 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18164 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61780 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17058 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74425 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67910 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12633 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16309 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62222 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12673 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59366 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62248 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/10209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89689 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10459 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43855 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15886 "Found 5 new JSC stress test failures: wasm.yaml/wasm/function-references/br_on_null.js.wasm-no-cjit, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/tail-call-js-inline.js.wasm-slow-memory, wasm.yaml/wasm/stress/tail-call.js.wasm-eager, wasm.yaml/wasm/v8/many-parameters.js.wasm-eager-jettison (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/46123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44671 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntTierUpCounter.h:
```diff
@@ -61,7 +61,8 @@ class LLIntTierUpCounter : public BaselineExecutionCounter {
     void optimizeAfterWarmUp()
     {
-        setNewThreshold(Options::thresholdForBBQOptimizeAfterWarmUp());
+        setNewThreshold(Options::useWasmLLInt() ? Options::thresholdForBBQOptimizeAfterWarmUp() : 0);
```
Let's change these numbers in Options.cpp when useWasmLLInt is set instead of changing it here.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntTierUpCounter.h:
```diff
@@ -63,7 +63,8 @@ class IPIntTierUpCounter : public BaselineExecutionCounter {
     {
-        setNewThreshold(Options::thresholdForBBQOptimizeAfterWarmUp());
+        setNewThreshold(Options::useWasmIPInt() ? Options::thresholdForBBQOptimizeAfterWarmUp() : 0);
+        ASSERT(Options::useWasmIPInt() || checkIfOptimizationThresholdReached());
```
Ditto.

kmiller68 committed:
BBQPlan should not be an EntryPlan
https://bugs.webkit.org/show_bug.cgi?id=280274
rdar://136594153

Reviewed by Yusuke Suzuki.

Right now BBQPlan is also an EntryPlan which means it needs to do all the
validation/module call graph steps required for entry. This was error prone
as e.g. BBQPlan with useWasmLLInt=0 would not report it's callsites to the
CalleeGroup thus would never would call OMG code.

This patch instead makes useWasmLLInt/useWasmIPInt=0 instantly trigger a
sync tier up to BBQ. This should simplify the code paths and make things
much simpler.

* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::BBQPlan):
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::compileFunction):
(JSC::Wasm::BBQPlan::prepareImpl): Deleted.
(JSC::Wasm::BBQPlan::didCompleteCompilation): Deleted.
(JSC::Wasm::BBQPlan::initializeCallees): Deleted.
(JSC::Wasm::BBQPlan::didReceiveFunctionData): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::jitCompileAndSetHeuristics):
* Source/JavaScriptCore/wasm/WasmIPIntTierUpCounter.h:
(JSC::Wasm::IPIntTierUpCounter::optimizeAfterWarmUp):
* Source/JavaScriptCore/wasm/WasmLLIntTierUpCounter.h:
(JSC::Wasm::LLIntTierUpCounter::optimizeAfterWarmUp):
* Source/JavaScriptCore/wasm/WasmModule.cpp:
(JSC::Wasm::Module::getOrCreateCalleeGroup):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::jitCompileAndSetHeuristics):
(JSC::LLInt::jitCompileSIMDFunction):

Canonical link: https://commits.webkit.org/284175@main
